pkg/utils/etcdutil/etcdutil.go

```diff
@@ -196,72 +210,237 @@ func EtcdKVPutWithTTL(ctx context.Context, c *clientv3.Client, key string, value
-	for _, u := range acUrls {
-		endpoints = append(endpoints, u.String())
-	}
+func newClient(tlsConfig *tls.Config, endpoints ...string) (*clientv3.Client, error) {
 	lgc := zap.NewProductionConfig()
 	lgc.Encoding = log.ZapEncodingName
-	autoSyncInterval := defaultAutoSyncInterval
-	dialKeepAliveTime := defaultDialKeepAliveTime
-	dialKeepAliveTimeout := defaultDialKeepAliveTimeout
-	failpoint.Inject("autoSyncInterval", func() {
-		autoSyncInterval = 10 * time.Millisecond
-	})
-	failpoint.Inject("closeKeepAliveCheck", func() {
-		autoSyncInterval = 0
-		dialKeepAliveTime = 0
-		dialKeepAliveTimeout = 0
-	})
 	client, err := clientv3.New(clientv3.Config{
 		Endpoints:            endpoints,
 		DialTimeout:          defaultEtcdClientTimeout,
-		AutoSyncInterval:     autoSyncInterval,
 		TLS:                  tlsConfig,
 		LogConfig:            &lgc,
-		DialKeepAliveTime:    dialKeepAliveTime,
-		DialKeepAliveTimeout: dialKeepAliveTimeout,
+		DialKeepAliveTime:    defaultDialKeepAliveTime,
+		DialKeepAliveTimeout: defaultDialKeepAliveTimeout,
 	})
-	if err == nil {
-		log.Info("create etcd v3 client", zap.Strings("endpoints", endpoints))
+	return client, err
+}
+
+// CreateEtcdClient creates etcd v3 client with detecting endpoints.
+func CreateEtcdClient(tlsConfig *tls.Config, acURLs []url.URL) (*clientv3.Client, error) {
+	urls := make([]string, 0, len(acURLs))
+	for _, u := range acURLs {
+		urls = append(urls, u.String())
 	}
+	client, err := newClient(tlsConfig, urls...)
+	if err != nil {
+		return nil, err
+	}
+
+	tickerInterval := defaultDialKeepAliveTime
+	failpoint.Inject("fastTick", func() {
+		tickerInterval = 100 * time.Millisecond
+	})
+	failpoint.Inject("closeTick", func() {
+		tickerInterval = 0
+	})
+	if tickerInterval == 0 {
+		return client, err
+	}
+
+	checker := &healthyChecker{
+		tlsConfig: tlsConfig,
+	}
+	eps := syncUrls(client)
+	checker.update(eps)
+
+	// Create a goroutine to check the health of etcd endpoints periodically.
+	go func(client *clientv3.Client) {
+		defer logutil.LogPanic()
+		ticker := time.NewTicker(tickerInterval)
+		defer ticker.Stop()
+		lastAvailable := time.Now()
+		for {
+			select {
+			case <-client.Ctx().Done():
+				log.Info("[etcd client] etcd client is closed, exit health check goroutine")
+				checker.Range(func(key, value interface{}) bool {
+					client := value.(*healthyClient)
+					client.Close()
+					return true
+				})
+				return
+			case <-ticker.C:
+				usedEps := client.Endpoints()
+				healthyEps := checker.patrol(client.Ctx())
+				if len(healthyEps) == 0 {
+					// when all endpoints are unhealthy, try to reset endpoints to update connect
+					// rather than delete them to avoid there is no any endpoint in client.
+					if time.Since(lastAvailable) > etcdServerDisconnectedTimeout {
+						log.Info("[etcd client] no available endpoint, try to reset endpoints", zap.Strings("last-endpoints", usedEps))
+						client.SetEndpoints([]string{}...)
```

Comment:
What is it used for?

Reply:
It will call [resolverGroup.SetEndpoints](https://github.com/etcd-io/etcd/blob/85b640cee793e25f3837c47200089d14a8392dc7/clientv3/client.go) , which will call [cc.NewAddress(addrs)](https://github.com/etcd-io/etcd/blob/85b640cee793e25f3837c47200089d14a8392dc7/clientv3/balancer/resolver/endpoint/endpoint.go#L89), which will call [updateResolverState](https://github.com/grpc/grpc-go/blob/f5b0812e6fe574d90da76b205e9eb51f6ddb1919/resolver_conn_wrapper.go#L183), which will call [updateClientConnState](https://github.com/grpc/grpc-go/blob/f5b0812e6fe574d90da76b205e9eb51f6ddb1919/clientconn.go#L638), which will call [balancer.HandleResolvedAddrs](https://github.com/grpc/grpc-go/blob/f5b0812e6fe574d90da76b205e9eb51f6ddb1919/balancer_conn_wrappers.go#L129)

It will close SubConn in https://github.com/etcd-io/etcd/blob/85b640cee793e25f3837c47200089d14a8392dc7/clientv3/balancer/balancer.go#L175 

If we don't close it manual, it will retry in [resetTransport](https://github.com/grpc/grpc-go/blob/f5b0812e6fe574d90da76b205e9eb51f6ddb1919/clientconn.go#L1082), and grpc use [exponential backoff strategy](https://github.com/grpc/grpc-go/blob/f5b0812e6fe574d90da76b205e9eb51f6ddb1919/internal/backoff/backoff.go#L43) by default, which means etcd client need to wait longer when etcd server is down.

In test, if etcd restart interval is 5min, the backoff interval in resetTransport maybe 2m16.425989119s

Reply:
Will it break the normal connection?

Reply:
It is triggered only when all endpoints are unavailable for a sustained period of time.

Reply:
https://github.com/tikv/pd/pull/6577/files#diff-bc08534cb4c71e0934cbd39c6cb3fb6a1d5bde0729cd2c81c484ff760ab00b87R317-R318 for this goroutine, are they unavailable?

Reply:
When all endpoints are unavailable, syncUrls also will be failed to list members, so it cannot help to recover as soon as possible in this case.

And client.SetEndpoints with empty and immediately set endpoints with used endpoints, subconn is removed and created again, it do not affect `syncUrls` when cluster recover.

In a word, L286 and L287 are used to linear retry rather than exponential.